demos/dev_mode_key_provisioning/src/aws_dev_mode_key_provisioning.c

```diff
@@ -43,88 +44,503 @@
+            xPrivateKeyTemplate.xCoefficient.ulValueLen = COEFFICIENT_LENGTH;
+            xPrivateKeyTemplate.xTokenObject.type = CKA_TOKEN;
+            xPrivateKeyTemplate.xTokenObject.pValue = &xTokenStorage;
+            xPrivateKeyTemplate.xTokenObject.ulValueLen = sizeof( xTokenStorage );
+        }
+
+        xResult = pxFunctionList->C_CreateObject( xSession,
+                                                  ( CK_ATTRIBUTE_PTR ) &xPrivateKeyTemplate,
+                                                  sizeof( PKCS11_PrivateRsaKeyTemplate_t ) / sizeof( CK_ATTRIBUTE ),
+                                                  pxObjectHandle );
+
+        if( NULL != pxModulus )
+        {
+            vPortFree( pxModulus );
+        }
+
+        if( NULL != pxE )
+        {
+            vPortFree( pxE );
+        }
+
+        if( NULL != pxD )
+        {
+            vPortFree( pxD );
+        }
+
+        if( NULL != pxPrime1 )
+        {
+            vPortFree( pxPrime1 );
+        }
+
+        if( NULL != pxPrime2 )
+        {
+            vPortFree( pxPrime2 );
+        }
+
+        if( NULL != pxExp1 )
+        {
+            vPortFree( pxExp1 );
+        }
+
+        if( NULL != pxExp2 )
+        {
+            vPortFree( pxExp2 );
+        }
+
+        if( NULL != pxCoefficient )
+        {
+            vPortFree( pxCoefficient );
+        }
     }
 
+    mbedtls_pk_free( &xMbedPkContext );
+
     return xResult;
 }
+
+
+CK_RV xProvisionPublicKey( CK_SESSION_HANDLE xSession,
+                           uint8_t * pucKey,
+                           size_t xKeyLength,
+                           CK_KEY_TYPE xPublicKeyType,
+                           uint8_t * pucPublicKeyLabel,
+                           CK_OBJECT_HANDLE_PTR pxPublicKeyHandle )
+{
+    CK_RV xResult;
+    CK_BBOOL xTrue = CK_TRUE;
+    CK_FUNCTION_LIST_PTR pxFunctionList;
+    CK_OBJECT_CLASS xClass = CKO_PUBLIC_KEY;
+
+    xResult = C_GetFunctionList( &pxFunctionList );
+
+    int lMbedResult = 0;
+    mbedtls_pk_context xMbedPkContext;
+
+    mbedtls_pk_init( &xMbedPkContext );
+    lMbedResult = mbedtls_pk_parse_key( &xMbedPkContext, pucKey, xKeyLength, NULL, 0 );
+
+    if( lMbedResult != 0 )
+    {
+        lMbedResult = mbedtls_pk_parse_public_key( &xMbedPkContext, pucKey, xKeyLength );
+    }
+
+    if( xPublicKeyType == CKK_RSA )
+    {
+        CK_BYTE xPublicExponent[] = { 0x01, 0x00, 0x01 };
+        CK_BYTE xModulus[ MODULUS_LENGTH + 1 ] = { 0 };
+
+        lMbedResult = mbedtls_rsa_export_raw( ( mbedtls_rsa_context * ) xMbedPkContext.pk_ctx,
+                                              ( unsigned char * ) &xModulus, MODULUS_LENGTH + 1,
+                                              NULL, 0,
+                                              NULL, 0,
+                                              NULL, 0,
+                                              NULL, 0 );
+        CK_ATTRIBUTE xPublicKeyTemplate[] =
```

Comment:
This is easy on the eyes.  Could the xPrivateKeyTemplate in the last function be populated the same way?